### PR TITLE
CAM: SelectLoop - Optimize

### PR DESCRIPTION
**Pull request by tarman3**
Added optimizations to decrease time processing for big amount of edges:
 
Example: [board_back.zip](https://github.com/user-attachments/files/24422841/board_back.zip)
Amount edges in shape: 1852
Amount selected edges in loop: 1092
Before: 30 sec
After: 1 sec

Screenshot_20260104_184402

### CODE REVIEW

**Comment by tarman3** [comment]
More optimizations

Model: [selectloop_test.zip](https://github.com/user-attachments/files/24432576/selectloop_test.zip)
Amount edges in shape: 5556
Amount selected edges in loop: 1092
Before: 2 min 20 sec
After: 3 sec

Screenshot_20260105_134444

**Comment by tarman3** [comment]
Attempt to optimize `horizontalFaceLoop()`

This part of code is very slow and take more than 2 minutes with test model above

https://github.com/FreeCAD/FreeCAD/blob/dfbdb2005cfb3822557a5a7945bacfd2333d331a/src/Mod/CAM/PathScripts/PathUtils.py#L262-L270

Does this code really needed?

**Comment by tarman3** [comment]
@sliptonic 
I found this topic from you in form: [Finding outer wire of irregular shapes](https://forum.freecad.org/viewtopic.php?t=17160)
So maybe my question bellow is related

I attempt to optimize method `PathScripts.PathUtils.horizontalFaceLoop()`
This part of code is very slow and take more than 2 minutes with test model above

https://github.com/FreeCAD/FreeCAD/blob/dfbdb2005cfb3822557a5a7945bacfd2333d331a/src/Mod/CAM/PathScripts/PathUtils.py#L262-L270

I want to exclude or try to optimize this part of code, but I do not have any example than this filter needed
Do you know something about the reasons to filter edges after `TechDraw.findShapeOutline()`?

Thank You

**Comment by tarman3** [comment]
@sliptonic 
Soft reminder
If possible, Please look at this https://github.com/FreeCAD/FreeCAD/pull/26644#issuecomment-3720755145

Thank  You

**Comment by sliptonic** [comment]
> @sliptonic I found this topic from you in form: [Finding outer wire of irregular shapes](https://forum.freecad.org/viewtopic.php?t=17160) So maybe my question bellow is related
> 
> I attempt to optimize method `PathScripts.PathUtils.horizontalFaceLoop()` This part of code is very slow and take more than 2 minutes with test model above
> 
> https://github.com/FreeCAD/FreeCAD/blob/dfbdb2005cfb3822557a5a7945bacfd2333d331a/src/Mod/CAM/PathScripts/PathUtils.py#L262-L270
> 
> I want to exclude or try to optimize this part of code, but I do not have any example than this filter needed Do you know something about the reasons to filter edges after `TechDraw.findShapeOutline()`?
> 
> Thank You

I'm sorry but my memory is failing me here.
I only recall that there were (are) a lot of corner cases and this was implemented before I was regularly writing unit tests.
If we had those unit tests now, refactoring this would be much easier.  
If you find test cases, let's start building a test suite.

**Comment by tarman3** [comment]
Maybe behavior of `TechDraw.findShapeOutline()` changed
I did not meet any cases in which this filter do something useful
It only spend time, a lot of time with selection of thousands edges

I prefer to disable this filter and return to this question when (if) someone meet this special case

Thank you

**Comment by sliptonic** [comment]
ok.  Maybe comment it out and leave a comment link to this PR

**Comment by tarman3** [comment]
Not sure understand correctly
Do You mean ask this question in forum?

**Comment by sliptonic** [comment]
No.  Just disable the filter by commenting out the code.  Add an additional comment, pointing to this issue (github).
So if someone in the future finds an edge case, traces it back to that bit of code, they can find the link back to this discussion and maybe write the necessary unit tests.

**Comment by tarman3** [comment]
Sounds good

File: src/Mod/CAM/PathScripts/PathUtils.py [hunk 1]
```diff
@@ -93,111 +93,148 @@ def segments(poly):
...
+
+    getSortedClusters() return edges without changes,
+    but can be unstable and freezes with some shapes"""
+
+    badClusters = Part.sortEdges(candidates)
+
+    # most simple case
+    if len(badClusters) == 1:
+        wire = Part.Wire(badClusters[0])
+        if len(wire.Edges) == len(candidates) and wire.isClosed:
```
Comment (sliptonic) [inline src/Mod/CAM/PathScripts/PathUtils.py]: Inconsistent usage of isClosed vs isClosed().

Line 163 uses wire.isClosed (property access), while line 190 and 198 use wire.isClosed() (method call). In FreeCAD's Part module, Wire.isClosed() is typically a method. If isClosed is a property, line 163 will always evaluate to truthy (the method object itself).

Verify which form is correct and make usage consistent.

FreeCAD Part.Wire isClosed method or property
Also applies to: 189-191

File: src/Mod/CAM/PathScripts/PathUtils.py [hunk 2]
```diff
@@ -209,21 +246,17 @@ def tangentEdgeLoop(obj, edge):
...
-        # only if found tangent edges
+    if len(loop) > 2 and edge2.hashCode() in [e.hashCode() for e in loop]:
         return loop
 
     return None
 
 
-def horizontalFaceLoop(obj, face, faceList=None):
+def horizontalFaceLoop(obj, faces, subNames):
     """horizontalFaceLoop(obj, face, faceList=None) ... returns a list of face names which form the walls of a vertical hole face is a part of.
```
Comment (sliptonic) [inline src/Mod/CAM/PathScripts/PathUtils.py]: docstring doesn't match function signature.

File: src/Mod/CAM/PathScripts/PathUtils.py [hunk 3]
```diff
@@ -93,111 +93,148 @@ def segments(poly):
...
-    if len(loops) > 0:
-        return loops[0].Edges
 
-    # Trying to find edges in loop without wires from shape
+    # Trying to find edges in horizontal wires of shape
+    for wire in obj.Shape.Wires:
+        if not Path.Geom.isRoughly(wire.BoundBox.ZLength, 0):
+            # skip not horizontal wire
+            continue
+        if any(edge.hashCode() == ehash for edge in wire.Edges):
```
Comment (sliptonic) [inline src/Mod/CAM/PathScripts/PathUtils.py]: The loop variable edge shadows the function parameter edge. The comparison edge.hashCode() == ehash will compare each wire edge against ehash (which was computed from the original parameter), but after the loop completes, edge refers to the last edge of the last wire, not the original parameter.

This works accidentally because ehash was captured before the loop, but the code is confusing and fragile.

     for wire in obj.Shape.Wires:
         if not Path.Geom.isRoughly(wire.BoundBox.ZLength, 0):
             # skip not horizontal wire
             continue
-        if any(edge.hashCode() == ehash for edge in wire.Edges):
+        if any(e.hashCode() == ehash for e in wire.Edges):
             return wire.Edges

**Review by sliptonic — COMMENTED** [review]
testing code rabbit AI code review.  
These seem like valid issues

File: src/Mod/CAM/PathScripts/PathUtils.py (hunk 1 above)
Comment (tarman3) [reply src/Mod/CAM/PathScripts/PathUtils.py]: Yes. This is typo
Fixed

File: src/Mod/CAM/PathScripts/PathUtils.py (hunk 2 above)
Comment (tarman3) [reply src/Mod/CAM/PathScripts/PathUtils.py]: Fixed

File: src/Mod/CAM/PathScripts/PathUtils.py (hunk 3 above)
Comment (tarman3) [reply src/Mod/CAM/PathScripts/PathUtils.py]: I agree with changes, but only for code readability

AI is wrong here
All what is happening inside `any()` does not change anything outside of `any()`

```
>>> l = [0, 1, 2, 3, 4, 5]
>>> a = 10
>>> any(a == 5 for a in l)
True
>>> a
10
```

File: src/Mod/CAM/PathScripts/PathUtils.py [hunk 4]
```diff
@@ -93,111 +93,148 @@ def segments(poly):
...
+                continue
+
+            # edge and hash changed, so try to find similar edge
+            for badEdge in badCluster:
+                if Path.Geom.edgesSimilar(badEdge, edge):
+                    goodClusters[-1].append(edge)
+                    candidatesCopy.remove(edge)
+                    break
+
+    if set([len(bc) for bc in badClusters]) == set([len(gc) for gc in goodClusters]):
```
Comment (sliptonic) [inline src/Mod/CAM/PathScripts/PathUtils.py]: The comparison set([len(bc) for bc in badClusters]) == set([len(gc) for gc in goodClusters]) compares unique length values, not the multiset of lengths. For example, if badClusters has lengths [2, 2, 3] and goodClusters has lengths [2, 3], the sets {2, 3} would be equal even though cluster counts differ.

Consider comparing sorted lists or using collections.Counter:

-    if set([len(bc) for bc in badClusters]) == set([len(gc) for gc in goodClusters]):
+    if sorted(len(bc) for bc in badClusters) == sorted(len(gc) for gc in goodClusters):

Comment (tarman3) [reply src/Mod/CAM/PathScripts/PathUtils.py]: Fixed
Thank you

**Comment by tarman3** [comment]
Changed to Draft

Created issue
- #28990

To not add unreliable crutches, better to solve issue in `Part.getSortedClusters()`

**Comment by tarman3** [comment]
Created pull request to fix `Part.getSortedClusters()`
- #29111

Wrote in the forum about this changes and asked for tests
https://forum.freecad.org/viewtopic.php?p=882234#p882234

**Comment by tarman3** [comment]
**Edited:**
Method `Part.getSortedClusters()` already present in command `SelectLoop`
So no sense to wait PR #29111
